src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs

```diff
@@ -56,9 +62,22 @@ internal abstract class PEAssemblyBuilderBase : PEModuleBuilder, Cci.IAssemblyRe
 
         public override ISourceAssemblySymbolInternal SourceAssemblyOpt => _sourceAssembly;
 
-        internal override ImmutableArray<NamedTypeSymbol> GetAdditionalTopLevelTypes()
+        internal override ImmutableArray<NamedTypeSymbol> GetAdditionalTopLevelTypes(DiagnosticBag diagnostics)
         {
-            return _additionalTypes;
+            var builder = ArrayBuilder<NamedTypeSymbol>.GetInstance();
+            builder.AddRange(_additionalTypes);
+
+            CreateEmbeddedAttributesIfNeeded(diagnostics);
```

Comment:
`GetAdditionalTopLevelTypes` should return the same diagnostics each time it is called. My question was whether diagnostics should be returned from this method (which is called multiple times) or another method (perhaps `CreateEmbeddedAttributesIfNeeded`, separated out from `GetAdditionalTopLevelTypes`) that is called at most once per emit. Your current implementation, with a lazy diagnostics field seems reasonable.